README.md
## Static libraries for building RcppCWB on Windows

Following the model of R packages such as [openssl](https://github.com/jeroen/openssl/tree/master/tools), [libcurl](https://github.com/jeroen/curl) or [tesseract](https://github.com/ropensci/tesseract), building [RcppCWB](https://github.com/PolMine/RcppCWB) on Windows relies on pre-compiled static libraries for C code with dependencies that are hard to satisfy on Windows. Offering cross-compiled binaries for download is an approach to ensure the portability of R packages with C code that works nicely. The [rwinlib](https://github.com/rwinlib) collection of repositories is an established place to host static libraries. This repository is the equivalent for building RcppCWB on Windows.

The approach is as follows:

- Executing an R script (file tools/winlibs.R within the package) is triggered by 'src/Makevars.win' when installing the package. 
- The 'winlibs.R' script performs a download of static libraries necessary to compile C/C++ code from a GitHub repository. In the case of RcppCWB, the content of this repository is downloaded as a zip file.
- The 'PKG_LIBS' variable then includes a flag pointing to the static libraries that have been downloaded.

This repository includes three static libraries with CWB code that have been cross-compiled on a Linux machine for 32bit and 64bit R (lib/i386 and lib/x64 directory, respectively). Additional static libraries are included to ensure that relevant dependencies are present:

- [GLib](https://developer.gnome.org/glib/)
- [LibIntl](http://gnuwin32.sourceforge.net/packages/libintl.htm)
- [PCRE](https://www.pcre.org)

The CWB static libraries have been cross-compiled using Docker containers. The respective Dockerfiles are part of this repository.

- The Dockerfile in docker/gcc4.8 uses a Ubuntu 14.04 Docker container and GCC 4.8 is used for cross-compilation. This is historically the approach that runs smoothly, with limited side-effects. A bash file will run through the process and yield both 32 and 64 bit versions of the static libraries. These are copied into the repository.
- The Dockerfile in docker/gcc11.2 can be used to check errors and warnings that result from the most recent version of GCC.

A note on branches:

- Starting with RcppCWB v0.4.4.9003, static libraries are retrieved from the branch named after the version number of the CWB version to build.
- The branch gcc11 is there for historic reasons, because it is the branch retrieved by the CRAN release of RcppCWB.
- The main branch of this repository is not actively used.





NEWS.md
[2023-03-20] Updated Dockerfile to work with pcre2 on new brach v3.4.33-patched
[2022-01-26] New Dockerfile to cross-compile CWB static libraries in Fedora container with GCC11.2: To learn about warnings/errors thrown by recent compiler.
[2022-01-20] Branch v3.4.33 created for static libraries for CWB v3.4.33. New approach to use CWB Makefiles. 2-step procedure with two Dockerfiles.
[2021-07-05] The ./buildenv/Dockerfile has been reworked entirely and will now yield libcl.a, libcqp.a and libcwb.a (32 and 64 bit)
[2021-07-06] Inserted ARG RCPPCWB_BUILD into ./buildenv/Dockerfile to trigger fresh clone of RcppCWB if the package is updated



include/cwb/cl/storage.h
/* 
 *  IMS Open Corpus Workbench (CWB)
 *  Copyright (C) 1993-2006 by IMS, University of Stuttgart
 *  Copyright (C) 2007-     by the respective contributers (see file AUTHORS)
 * 
 *  This program is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by the
 *  Free Software Foundation; either version 2, or (at your option) any later
 *  version.
 * 
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 *  Public License for more details (in the file "COPYING", or available via
 *  WWW at http://www.gnu.org/copyleft/gpl.html).
 */

#ifndef __storage_h
#define __storage_h

#include <sys/types.h>

#include "globals.h"


/* data allocation methods */

#define UNALLOCATED 0
#define MMAPPED  1    /**< Flag: use mmap() to allocate memory */
#define MALLOCED 2    /**< Flag: use malloc() to allocate memory */
#define PAGED    3


#define SIZE_BIT   0
#define SIZE_BYTE  sizeof(char)
#define SIZE_SHINT sizeof(short)
#define SIZE_INT   sizeof(int)
#define SIZE_LONG  sizeof(long)


/**
 * The MemBlob object.
 *
 * This object, unsurprisingly, represents a blob of memory.
 */
typedef struct TMblob {
  size_t size;                  /**< the number of allocated bytes */

  int item_size;                /**< the size of one item */
  unsigned int nr_items;        /**< the number of items represented */

  int *data;                    /**< pointer to the data */
  int allocation_method;        /**< the allocation method */

  int writeable;                /**< can we write to the data? */
  int changed;                  /**< needs update? (not yet in use) */

  /* fields for paged memory -- not yet used */
  char *fname;
  off_t fsize, offset;
} MemBlob;



/* ---------------------------------------------------------------------- */



void NwriteInt(int val, FILE *fd);
void NreadInt(int *val, FILE *fd);



void NwriteInts(int *vals, int nr_vals, FILE *fd);
void NreadInts(int *vals, int nr_vals, FILE *fd);


/* ---------------------------------------------------------------------- */



void mfree(MemBlob *blob);

void init_mblob(MemBlob *blob);

int alloc_mblob(MemBlob *blob, int nr_items, int item_size, int clear_blob);




/* ================================================================ FILE IO */



int read_file_into_blob(char *filename, 
                        int allocation_method,
                        int item_size,
                        MemBlob *blob);




int write_file_from_blob(char *filename, 
                         MemBlob *blob,
                         int convert_to_nbo);

/* ======================================================= ACCESS FUNCTIONS */

/* not yet implemented */

/* ==================================================== LOW LEVEL FUNCTIONS */

void * mmapfile(char *filename, size_t *len_ptr, char *mode);

void * mallocfile(char *filename, size_t *len_ptr, char *mode);

/* a new-style API for MemBlobs */
/* NB argument orders for the read/write functions are wrong... */
#define memblob_read_from_file read_file_into_blob
#define memblob_write_to_file write_file_from_blob
#define memblob_free mfree
#define memblob_clear init_mblob
#define memblob_allocate alloc_mblob


#endif


include/cwb/cl/registry.tab.h
/* A Bison parser, made by GNU Bison 3.0.2.  */

/* Bison interface for Yacc-like parsers in C

   Copyright (C) 1984, 1989-1990, 2000-2013 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* As a special exception, you may create a larger work that contains
   part or all of the Bison parser skeleton and distribute that work
   under terms of your choice, so long as that work isn't itself a
   parser generator using the skeleton or a modified version thereof
   as a parser skeleton.  Alternatively, if you modify or redistribute
   the parser skeleton itself, you may (at your option) remove this
   special exception, which will cause the skeleton and the resulting
   Bison output files to be licensed under the GNU General Public
   License without this special exception.

   This special exception was added by the Free Software Foundation in
   version 2.2 of Bison.  */

#ifndef YY_CREG_REGISTRY_TAB_H_INCLUDED
# define YY_CREG_REGISTRY_TAB_H_INCLUDED
/* Debug traces.  */
#ifndef YYDEBUG
# define YYDEBUG 1
#endif
#if YYDEBUG
extern int cregdebug;
#endif

/* Token type.  */
#ifndef YYTOKENTYPE
# define YYTOKENTYPE
  enum yytokentype
  {
    NAME_SYM = 258,
    ID_SYM = 259,
    INFO_SYM = 260,
    HOME_SYM = 261,
    ATTRIBUTE_SYM = 262,
    DIR_SYM = 263,
    CORPUS_SYM = 264,
    REVCORP_SYM = 265,
    REVCIDX_SYM = 266,
    FREQS_SYM = 267,
    LEXICON_SYM = 268,
    LEXIDX_SYM = 269,
    LEXSRT_SYM = 270,
    STRUCTURE_SYM = 271,
    ALIGNED_SYM = 272,
    DYNAMIC_SYM = 273,
    DOTS_SYM = 274,
    IGNORE_SYM = 275,
    ADMIN_SYM = 276,
    ACCESS_SYM = 277,
    USER_SYM = 278,
    GROUP_SYM = 279,
    ASSERT_SYM = 280,
    HOST_SYM = 281,
    PROPERTY_SYM = 282,
    IDENTIFIER = 283,
    STRING = 284,
    NUMBER = 285
  };
#endif

/* Value type.  */
#if ! defined YYSTYPE && ! defined YYSTYPE_IS_DECLARED
typedef union YYSTYPE YYSTYPE;
union YYSTYPE
{
#line 69 "registry.y" /* yacc.c:1909  */

  char    *strval;
  int      ival;
  void    *args;
  void    *attr;

  IDList   idlist;

  struct {
    int status;
    char *path;
  } storage;

#line 99 "registry.tab.h" /* yacc.c:1909  */
};
# define YYSTYPE_IS_TRIVIAL 1
# define YYSTYPE_IS_DECLARED 1
#endif


extern YYSTYPE creglval;

int cregparse (void);

#endif /* !YY_CREG_REGISTRY_TAB_H_INCLUDED  */


include/cwb/cl/class-mapping.h
/* 
 *  IMS Open Corpus Workbench (CWB)
 *  Copyright (C) 1993-2006 by IMS, University of Stuttgart
 *  Copyright (C) 2007-     by the respective contributers (see file AUTHORS)
 * 
 *  This program is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by the
 *  Free Software Foundation; either version 2, or (at your option) any later
 *  version.
 * 
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 *  Public License for more details (in the file "COPYING", or available via
 *  WWW at http://www.gnu.org/copyleft/gpl.html).
 */

#ifndef _MAPPING_H_
#define _MAPPING_H_

#include "globals.h"
#include "corpus.h"
#include "attributes.h"



/**
 * The SingleMapping object.
 *
 * This object represents a simple many-to-one mapping between one "class" and
 * a number of different tokens (represented by integer ID codes).
 *
 * The tokens are values that occur on a particular attribute in a particular
 * corpus. (Note: positional attributes, not other types!)
 * [?? -- is this correct? -- AH]
 */
typedef struct _single_mapping {
  char *class_name;             /**< this object's class name */
  int nr_tokens;                /**< the number of tokens which this object owns */
  int *tokens;                  /**< the tokens themselves */
} SingleMappingRecord, 
    *SingleMapping;

/**
 * The Mapping object.
 *
 * A Mapping consists of a bundle of SingleMappings. Each SingleMapping represents
 * a class within the mapping, and each class maps to a number of different tokens
 * on a particular p-attribute in a particular corpus.
 *
 * Classes are mutually exclusive, i.e. a single token cannot be a member
 * of more than one class in the same Mapping.
 *
 * NB: "Mapping" is the pointer (class); MappingRecord is the underlying structure.
 *
 * @see SingleMapping
 */
typedef struct _mapping {
  Corpus *corpus;               /**< the corpus this Mapping is valid in */
  Attribute *attribute;         /**< the attribute this object is valid for */
  char *mapping_name;           /**< this Mapping's name */
  int nr_classes;               /**< the number of single mappings currently in this Mapping */
  SingleMappingRecord *classes; /**< the mappings themselves */
} MappingRecord, *Mapping;






/* -------------------- create/destruct mappings */

Mapping
read_mapping(Corpus *corpus,
             char *attr_name,
             char *file_name,
             char **error_string);


int
drop_mapping(Mapping *map);

void
print_mapping(Mapping map);

/* -------------------- token -> class */

SingleMapping
map_token_to_class(Mapping map, 
                   char *token);

int
map_token_to_class_number(Mapping map, 
                          char *token);

int
map_id_to_class_number(Mapping map, 
                       int id);

/* -------------------- class -> {tokens} */

int *
map_class_to_tokens(SingleMapping map,
                    int *nr_tokens);


/* -------------------- utils */

int
number_of_classes(Mapping map);

SingleMapping
find_mapping(Mapping map, char *name);

int
number_of_tokens(SingleMapping map);

/* -------------------- predicates */

int
member_of_class_s(Mapping map, 
                  SingleMapping class_, 
                  char *token);

int
member_of_class_i(Mapping map, 
                  SingleMapping class_, 
                  int id);

#endif


include/cwb/cqp/parser.tab.h
/* A Bison parser, made by GNU Bison 3.0.2.  */

/* Bison interface for Yacc-like parsers in C

   Copyright (C) 1984, 1989-1990, 2000-2013 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* As a special exception, you may create a larger work that contains
   part or all of the Bison parser skeleton and distribute that work
   under terms of your choice, so long as that work isn't itself a
   parser generator using the skeleton or a modified version thereof
   as a parser skeleton.  Alternatively, if you modify or redistribute
   the parser skeleton itself, you may (at your option) remove this
   special exception, which will cause the skeleton and the resulting
   Bison output files to be licensed under the GNU General Public
   License without this special exception.

   This special exception was added by the Free Software Foundation in
   version 2.2 of Bison.  */

#ifndef YY_YY_PARSER_TAB_H_INCLUDED
# define YY_YY_PARSER_TAB_H_INCLUDED
/* Debug traces.  */
#ifndef YYDEBUG
# define YYDEBUG 1
#endif
#if YYDEBUG
extern int yydebug;
#endif

/* Token type.  */
#ifndef YYTOKENTYPE
# define YYTOKENTYPE
  enum yytokentype
  {
    ID = 258,
    QID = 259,
    NQRID = 260,
    LABEL = 261,
    STRING = 262,
    FLAG = 263,
    TAGSTART = 264,
    TAGEND = 265,
    VARIABLE = 266,
    IPAddress = 267,
    IPSubnet = 268,
    INTEGER = 269,
    DOUBLEFLOAT = 270,
    FIELD = 271,
    FIELDLABEL = 272,
    ANCHORTAG = 273,
    ANCHORENDTAG = 274,
    SEARCH_STRATEGY = 275,
    TAB_SYM = 276,
    CAT_SYM = 277,
    DEFINE_SYM = 278,
    DIFF_SYM = 279,
    DISCARD_SYM = 280,
    EXPAND_SYM = 281,
    EXIT_SYM = 282,
    FLAT_SYM = 283,
    INTER_SYM = 284,
    JOIN_SYM = 285,
    SUBSET_SYM = 286,
    LEFT_SYM = 287,
    RIGHT_SYM = 288,
    SAVE_SYM = 289,
    SCATTER_SYM = 290,
    SHOW_SYM = 291,
    CD_SYM = 292,
    TO_SYM = 293,
    WITHIN_SYM = 294,
    SET_SYM = 295,
    EXEC_SYM = 296,
    CUT_SYM = 297,
    OCCURS_SYM = 298,
    INFO_SYM = 299,
    GROUP_SYM = 300,
    WHERE_SYM = 301,
    ESCAPE_SYM = 302,
    MEET_SYM = 303,
    UNION_SYM = 304,
    MU_SYM = 305,
    SORT_SYM = 306,
    COUNT_SYM = 307,
    ASC_SYM = 308,
    DESC_SYM = 309,
    REVERSE_SYM = 310,
    BY_SYM = 311,
    FOREACH_SYM = 312,
    ON_SYM = 313,
    YES_SYM = 314,
    OFF_SYM = 315,
    NO_SYM = 316,
    SLEEP_SYM = 317,
    REDUCE_SYM = 318,
    MAXIMAL_SYM = 319,
    WITH_SYM = 320,
    WITHOUT_SYM = 321,
    DELETE_SYM = 322,
    SIZE_SYM = 323,
    DUMP_SYM = 324,
    UNDUMP_SYM = 325,
    TABULATE_SYM = 326,
    NOT_SYM = 327,
    CONTAINS_SYM = 328,
    MATCHES_SYM = 329,
    GCDEL = 330,
    APPEND = 331,
    LET = 332,
    GET = 333,
    NEQ = 334,
    IMPLIES = 335,
    RE_PAREN = 336,
    EOL_SYM = 337,
    ELLIPSIS = 338,
    MATCHALL = 339,
    LCSTART = 340,
    LCEND = 341,
    LCMATCHALL = 342,
    EXTENSION = 343,
    PLUSEQ = 344,
    MINUSEQ = 345,
    UNLOCK_SYM = 346,
    USER_SYM = 347,
    HOST_SYM = 348,
    UNDEFINED_MACRO = 349,
    MACRO_SYM = 350,
    RANDOMIZE_SYM = 351,
    FROM_SYM = 352,
    INCLUSIVE_SYM = 353,
    EXCLUSIVE_SYM = 354,
    NULL_SYM = 355
  };
#endif

/* Value type.  */
#if ! defined YYSTYPE && ! defined YYSTYPE_IS_DECLARED
typedef union YYSTYPE YYSTYPE;
union YYSTYPE
{
#line 120 "parser.y" /* yacc.c:1909  */

  Evaltree           evalt;
  Constrainttree     boolt;
  enum b_ops         boolo;
  int                ival;
  double             fval;
  int                index;
  char              *strval;
  CorpusList        *cl;

  struct {
    int a, b;
  } intpair;

  Context            context;
  ActualParamList   *apl;

  enum ctxtdir       direction;

  struct Redir       redir;

  struct InputRedir  in_redir;

  struct {
    int ok;
    int ival;
    char *cval;
  }                  varval;

  struct {
    FieldType field;
    int inclusive;
  }                  base;

  struct {
    char *variableValue;
    char operator;
  }                  varsetting;

  struct {
    int mindist;
    int maxdist;
  }                  Distance;

  struct {
    FieldType anchor;
    int offset;
  }                  Anchor;

  struct {
    FieldType anchor1;
    int offset1;
    FieldType anchor2;
    int offset2;
  }                  AnchorPair;

  struct {
    char *name;
    int flags;
  }                  AttributeSpecification;

  RangeSetOp         rngsetop;

  SortClause         sortclause;

  FieldType          field;

  SearchStrategy     search_strategy;

  TabulationItem     tabulation_item;

#line 227 "parser.tab.h" /* yacc.c:1909  */
};
# define YYSTYPE_IS_TRIVIAL 1
# define YYSTYPE_IS_DECLARED 1
#endif


extern YYSTYPE yylval;

int yyparse (void);

#endif /* !YY_YY_PARSER_TAB_H_INCLUDED  */
